modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -279,107 +281,124 @@
-                          {:name column-name
-                           :table-name table-name
-                           :table-schema dataset-id
-                           :database-type database-type
-                           :base-type base-type
-                           :database-partitioned partitioned?
-                           :database-position database-position}
-                          nil
-                          database-position)]
-                  ;; _PARTITIONDATE does not appear so add it in if we see _PARTITIONTIME
-                  (= column-name partitioned-time-field-name)
-                  (conj {:name partitioned-date-field-name
-                         :table-name table-name
-                         :table-schema dataset-id
-                         :database-type "DATE"
-                         :base-type :type/Date
-                         :database-position (inc database-position)
-                         :database-partitioned true}))))
-            named-rows)))
+  (assert (seq table-names))
+  (let [named-rows-query {:select [:table_name :column_name :data_type :ordinal_position
+                                   [[:= :is_partitioning_column "YES"] :partitioned]]
+                          :from [[(information-schema-table project-id dataset-id "COLUMNS") :c]]
+                          :where [:in :table_name table-names]
+                          :order-by [:table_name :ordinal_position :column_name]}
+        named-rows (try (query-honeysql driver database named-rows-query)
+                        (catch Throwable e
+                          (log/warnf e "error in describe-fields for dataset: %s" dataset-id)))
+        nested-column-lookup (get-nested-columns-for-tables driver database project-id dataset-id table-names)]
+    (eduction
+     (x/by-key :table_name (x/into []))
+     (mapcat #(describe-dataset-rows nested-column-lookup dataset-id %))
+     named-rows)))
+
+(def ^:private ^:const num-table-partitions
+  "Number of tables to batch for describe-fields. Too low and we'll do too many queries, which is slow.
+   Too high and we'll hold too many fields of a dataset in memory, which risks causing OOMs."
+  1024)
+
+(defn- list-table-names [driver database project-id dataset-id]
+  (try
+    (eduction (map :table_name)
+              (query-honeysql driver database
+                              {:select [:table_name]
+                               :from [[(information-schema-table project-id dataset-id "TABLES") :t]]}))
+    (catch Throwable e
+      (log/warnf e "error in list-table-names for dataset: %s" dataset-id))))
 
 (defmethod driver/describe-fields :bigquery-cloud-sdk
   [driver database & {:keys [schema-names table-names]}]
   (let [project-id (get-project-id (:details database))
-        dataset-ids (or schema-names
-                        (list-datasets (:details database)))]
-    (sort-by
-     (juxt :table-schema :table-name :database-position :name)
```

Comment:
fields are now returned by "construction" ordered by `table-schema` and `table-name` (see the order-by in `named-rows-query`)